flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/TimestampsAndWatermarksOperator.java

```diff
@@ -0,0 +1,227 @@
+
+		watermarkInterval = getExecutionConfig().getAutoWatermarkInterval();
+		if (watermarkInterval > 0) {
+			final long now = getProcessingTimeService().getCurrentProcessingTime();
+			getProcessingTimeService().registerTimer(now + watermarkInterval, this);
+		}
+	}
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		IN value = element.getValue();
+		final long previousTimestamp = element.hasTimestamp() ?
+			element.getTimestamp() : Long.MIN_VALUE;
+
+		resuableInput.setField(0, previousTimestamp);
+		resuableInput.setField(1, value);
+
+		runnerInputSerializer.serialize(resuableInput, baosWrapper);
+		pythonFunctionRunner.process(baos.toByteArray());
+		baos.reset();
+		elementCount++;
+		checkInvokeFinishBundleByCount();
+		emitResults();
+	}
+
+	@Override
+	public void emitResult(Tuple2<byte[], Integer> resultTuple) throws Exception {
+		byte[] rawResult = resultTuple.f0;
+		int length = resultTuple.f1;
+		bais.setBuffer(rawResult, 0, length);
+		Row runnerOutput = (Row) runnerOutputSerializer.deserialize(baisWrapper);
+		long newTimestamp = (Long) runnerOutput.getField(0);
+		IN originalData = (IN) runnerOutput.getField(1);
+		reusableStreamRecord.replace(originalData, newTimestamp);
+		output.collect(reusableStreamRecord);
+		watermarkGenerator.onEvent(originalData, newTimestamp, watermarkOutput);
+
+	}
+
+	@Override
+	public PythonFunctionRunner createPythonFunctionRunner() throws Exception {
+		return new BeamDataStreamPythonFunctionRunner(
+			getRuntimeContext().getTaskName(),
+			createPythonEnvironmentManager(),
+			runnerInputTypeInfo,
+			runnerOutputTypeInfo,
+			DATA_STREAM_STATELESS_PYTHON_FUNCTION_URN,
+			PythonOperatorUtils.getUserDefinedDataStreamFunctionProto(pythonFunctionInfo, getRuntimeContext(), Collections.EMPTY_MAP),
+			DATA_STREAM_MAP_FUNCTION_CODER_URN,
+			jobOptions,
+			getFlinkMetricContainer(),
+			null,
+			null,
+			getContainingTask().getEnvironment().getMemoryManager(),
+			getOperatorConfig().getManagedMemoryFractionOperatorUseCaseOfSlot(
+				ManagedMemoryUseCase.PYTHON,
+				getContainingTask().getEnvironment().getTaskManagerInfo().getConfiguration(),
+				getContainingTask().getEnvironment().getUserCodeClassLoader().asClassLoader())
+		);
+	}
+
+	@Override
+	public void onProcessingTime(long timestamp) {
+		watermarkGenerator.onPeriodicEmit(watermarkOutput);
+		final long now = getProcessingTimeService().getCurrentProcessingTime();
+		getProcessingTimeService().registerTimer(now + watermarkInterval, this);
+	}
+
+	@Override
+	public void processWatermark(org.apache.flink.streaming.api.watermark.Watermark mark) {
+		if (mark.getTimestamp() == Long.MAX_VALUE) {
+			watermarkOutput.emitWatermark(Watermark.MAX_WATERMARK);
+		}
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		watermarkGenerator.onPeriodicEmit(watermarkOutput);
+	}
+
+	private static final class WatermarkEmitter implements WatermarkOutput {
```

Comment:
could reuse the WatermarkEmitter defined in org.apache.flink.streaming.runtime.operators.WatermarkEmitter.